HARK/dcegm.py

```diff
@@ -8,428 +8,298 @@
+            # If there are no valid crossings, return empty arrays.
+            points = np.zeros((0, 2), dtype=np.float64)
+            segments = np.zeros((0, 2), dtype=np.int64)
+            return points, segments
+
+        else:
 
-        return xing_points, segments
+            # Otherwise, subset valid crossings
+            segments = segments[valid, :]
+            switch_interv = switch_interv[valid, :]
+            left_y = left_y[valid, :]
+            right_y = right_y[valid, :]
 
+            # Find crossing points.
+            xing_points = [
+                calc_linear_crossing(switch_interv[i, :], left_y[i, :], right_y[i, :])
+                for i in range(segments.shape[0])
+            ]
 
-def calc_prim_kink(mGrid, vTGrids, choices):
+            xing_array = np.asarray(xing_points)
+
+            return xing_array, segments
+
+
+def calc_nondecreasing_segments(x, y):
     """
+    Given a sequence of (x,y) points, this function finds the start and end
+    indices of its largest non-decreasing segments.
+    
+    A non-decreasing segment is a sub-sequence of points
+    {(x_0, y_0),...,(x_n,y_n)} such that for all 0 <= i,j <= n,
+    If j>=i then x_j >= x_i and y_j >= y_i
+
     Parameters
     ----------
-    mGrid : np.array
-        Common m grid
-    vTGrids : [np.array], length  = # choices, each element has length = len(mGrid)
-        value functions evaluated on the common m grid.
-    choices : [np.array], length  = # choices, each element has length = len(mGrid)
-        Optimal choices. In the form of choice probability vectors that must
-        be degenerate
+    x : 1D np.array of floats
+        x coordinates of the sequence of points.
+    y : 1D np.array of floats
+        y coordinates of the sequence of points.
 
     Returns
     -------
-    kinks: [(mCoord, vTCoor)]
-        list of kink points
-    segments: [(left, right)]
-        List of the same length as kinks, where each element is a tuple
-        indicating which segments are optimal on each side of the kink.
-    """
+    starts : 1D np.array of ints
+        Indices where a new non-decreasing segment starts.
+    ends : 1D np.array of ints
+        Indices where a non-decreasing segment ends.
 
-    # Construct a vector with the optimal choice at each m point
-    optChoice = np.zeros_like(mGrid, dtype=int)
-    for i in range(len(vTGrids)):
-        idx = choices[i] == 1
-        optChoice[idx] = i
+    """
 
-    return calc_cross_points(mGrid, vTGrids.T, optChoice)
+    if len(x) == 0 or len(y) == 0 or len(y) != len(x):
+        raise Exception("x and y must be non-empty arrays of the same size.")
 
+    # Initialize
+    starts = [0]
+    ends = []
 
-def calc_segments(x, v):
-    """
-    Find index vectors `rise` and `fall` such that `rise` holds the indeces `i`
-    such that x[i+1]>x[i] and `fall` holds indeces `j` such that either
-    - x[j+1] < x[j] or,
-    - x[j]>x[j-1] and v[j]<v[j-1].
+    for i in range(1, len(x)):
```

Comment:
`for i in prange(1, len(x)):` for example

Reply:
Just pushed a new version, now njitted and without (explicit) loops